File: myonlinefm/src/main/java/com/sec/myonlinefm/updataUIListener/ObserverUIListenerManager.java
package com.sec.myonlinefm.updataUIListener;

import com.sec.myonlinefm.data.SearchType;
import com.sec.myonlinefm.data.Station;
import com.sec.myonlinefm.data.StationProgram;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Created by SRC-TJ-MM-BinYang on 2018/1/31.
 */

public class ObserverUIListenerManager implements SubjectUIListener {

    private static ObserverUIListenerManager observerUIManager;
    //观察者接口集合
    private List<ObserverUIListener> list = new ArrayList<>();

    /**
     * 单例
     */
    public static ObserverUIListenerManager getInstance(){
        if (null == observerUIManager){
            synchronized (ObserverUIListenerManager.class){
                if (null == observerUIManager){
                    observerUIManager = new ObserverUIListenerManager();
                }
            }
        }
        return observerUIManager;
    }

    @Override
    public void add(ObserverUIListener observerUIListener) {
        list.add(observerUIListener);
    }

    @Override
    public void notifyLiveRadioLocalObserver(List<Station> mStations, Map<Integer, List<StationProgram>> map) {
        for (ObserverUIListener observerUIListener : list){
            observerUIListener.observerLiveRadioLocalUIUpData(mStations, map);
        }
    }

    @Override
    public void notifyLiveRadioCenterObserver(List<Station> mStations, Map<Integer, List<StationProgram>> map) {
        for (ObserverUIListener observerUIListener : list){
            observerUIListener.observerLiveRadioCenterUIUpData(mStations, map);
        }
    }

    @Override
    public void notifyDifferentInfoObserver(List<Station> mStations, Map<Integer, List<StationProgram>> map) {
        for (ObserverUIListener observerUIListener : list){
            observerUIListener.observerDifferentInfoUIUpData(mStations, map);
        }
    }

    @Override
    public void notifyChannelLiveObserver(List<SearchType.ChannelLive> mSearchChannelLiveList) {
        for (ObserverUIListener observerUIListener : list){
            observerUIListener.observerChannelLiveUpData(mSearchChannelLiveList);
        }
    }

    @Override
    public void notifyProgramLiveObserver(List<SearchType.ProgramLive> mSearchProgramLiveList) {
        for (ObserverUIListener observerUIListener : list){
            observerUIListener.observerProgramLiveUpData(mSearchProgramLiveList);
        }
    }

    @Override
    public void notifyOneDayProgramUpData(List<StationProgram> mOneDayPrograms) {
        for (ObserverUIListener observerUIListener : list){
            observerUIListener.observerOneDayProgramUpData(mOneDayPrograms);
        }
    }

    @Override
    public void notifyConnectStatus(boolean isConnectNet) {
        for (ObserverUIListener observerUIListener : list){
            observerUIListener.observerConnectStatus(isConnectNet);
        }
    }

    @Override
    public void remove(ObserverUIListener observerUIListener) {
        if(list.contains(observerUIListener))
            list.remove(observerUIListener);
    }
}

File: myonlinefm/src/main/java/com/sec/myonlinefm/updataUIListener/ObserverUIListenerManager.kt
package com.sec.myonlinefm.updataUIListener

import com.sec.myonlinefm.data.SearchType
import com.sec.myonlinefm.data.Station
import com.sec.myonlinefm.data.StationProgram

/**
 * Created by SRC-TJ-MM-BinYang on 2018/1/31.
 */
class ObserverUIListenerManager : SubjectUIListener {
    //观察者接口集合
    private val list: MutableList<ObserverUIListener?> = ArrayList()
    override fun add(observerUIListener: ObserverUIListener?) {
        list.add(observerUIListener)
    }

    override fun notifyLiveRadioLocalObserver(mStations: MutableList<Station?>?, map: MutableMap<Int?, MutableList<StationProgram?>?>?) {
        for (observerUIListener in list) {
            observerUIListener!!.observerLiveRadioLocalUIUpData(mStations, map)
        }
    }

    override fun notifyLiveRadioCenterObserver(mStations: MutableList<Station?>?, map: MutableMap<Int?, MutableList<StationProgram?>?>?) {
        for (observerUIListener in list) {
            observerUIListener!!.observerLiveRadioCenterUIUpData(mStations, map)
        }
    }

    override fun notifyDifferentInfoObserver(mStations: MutableList<Station?>?, map: MutableMap<Int?, MutableList<StationProgram?>?>?) {
        for (observerUIListener in list) {
            observerUIListener!!.observerDifferentInfoUIUpData(mStations, map)
        }
    }

    override fun notifyChannelLiveObserver(mSearchChannelLiveList: MutableList<SearchType.ChannelLive?>?) {
        for (observerUIListener in list) {
            observerUIListener!!.observerChannelLiveUpData(mSearchChannelLiveList)
        }
    }

    override fun notifyProgramLiveObserver(mSearchProgramLiveList: MutableList<SearchType.ProgramLive?>?) {
        for (observerUIListener in list) {
            observerUIListener!!.observerProgramLiveUpData(mSearchProgramLiveList)
        }
    }

    override fun notifyOneDayProgramUpData(mOneDayPrograms: MutableList<StationProgram?>?) {
        for (observerUIListener in list) {
            observerUIListener!!.observerOneDayProgramUpData(mOneDayPrograms)
        }
    }

    override fun notifyConnectStatus(isConnectNet: Boolean) {
        for (observerUIListener in list) {
            observerUIListener!!.observerConnectStatus(isConnectNet)
        }
    }

    override fun remove(observerUIListener: ObserverUIListener?) {
        if (list.contains(observerUIListener)) list.remove(observerUIListener)
    }

    companion object {
        private var observerUIManager: ObserverUIListenerManager? = null

        /**
         * 单例
         */
        fun getInstance(): ObserverUIListenerManager? {
            if (null == observerUIManager) {
                synchronized(ObserverUIListenerManager::class.java) {
                    if (null == observerUIManager) {
                        observerUIManager = ObserverUIListenerManager()
                    }
                }
            }
            return observerUIManager
        }
    }
}